add callbacks before and after cluster operations. Refactoring

albertshau opened the pull request:
some common logic into TaskService to make sure the before and
after hooks are isolated to a single class.

albertshau committed:
add callbacks before and after cluster operations. Refactoring
some common logic into TaskService to make sure the before and
after hooks are isolated to a single class.

nitinmotgi commented on server/src/main/java/com/continuuity/loom/scheduler/callback/HttpPostClusterCallback.java:
```diff
@@ -0,0 +1,143 @@
+/**
+ * Executes before and after hooks by sending an HTTP POST request to some configurable endpoints, with the post body
+ * containing the cluster and job objects. By default the requests are sent before and after cluster creation and
+ * cluster deletion,
```
Is there cluster creation failure, cluster update and service start/stop ?

albertshau replied:
yes, these get called before and after (failure counts as after) any cluster operation including service stop,start,restart and reconfigure.  This particular callback I've defaulted to only do on create and delete.  Do you think it should default to every cluster operation?

albertshau replied:
hm maybe it would be better to have another method for failures instead of lumping it into the after call

albertshau committed:
Merge remote-tracking branch 'origin/develop' into feature/http-callbacks

Conflicts:
	server/src/main/java/com/continuuity/loom/scheduler/JobScheduler.java
	server/src/main/java/com/continuuity/loom/scheduler/task/TaskService.java
	server/src/test/java/com/continuuity/loom/scheduler/SchedulerTest.java

albertshau committed:
separate into start, success, and failure events instead of just
before and after.

albertshau committed:
update javadoc

poornachandra commented on server/src/main/java/com/continuuity/loom/scheduler/callback/HttpPostClusterCallback.java:
```diff
@@ -0,0 +1,163 @@
+        post.setEntity(new StringEntity(GSON.toJson(data)));
+        HttpResponse response = httpClient.execute(post);
+        // ignore what we get back, but make sure resources are released
+        HttpEntity entity = response.getEntity();
```
This call needs to be async so that the pipeline does not get delayed.

albertshau committed:
Merge remote-tracking branch 'origin/develop' into feature/http-callbacks

albertshau committed:
making the callbacks async, though the start callback for a
particular cluster still has to be finished before the tasks for
that cluster job are queued.

poornachandra commented on server/src/main/java/com/continuuity/loom/scheduler/callback/ClusterCallbackExecutor.java:
```diff
@@ -0,0 +1,101 @@
+        clusterCallback.onStart(data);
+      }
+    });
+    future.addListener(new Runnable() {
```
When the server goes down after calling the callback and comes back up, the future is lost and the job will not make any progress, right?

albertshau replied:
will use a queue like in other places for this.  Also means the callback should be idempotent, which will be documented

albertshau committed:
add nodes to callback post request, write callbacks to a queue
so that they are persisted across server failures, and adding
tests for the http post callback.

poornachandra commented on server/src/main/java/com/continuuity/loom/scheduler/callback/HttpPostClusterCallback.java:
```diff
@@ -0,0 +1,167 @@
+        // ignore what we get back, but make sure resources are released
+        HttpEntity entity = response.getEntity();
+        if (entity != null) {
+          entity.getContent().close();
```
It is generally a good practice to have close in a finally block.

albertshau replied:
this actually isn't needed after upgrading the client version and having post.releaseConnection in the finally.  will take it out.

poornachandra commented on server/src/main/java/com/continuuity/loom/scheduler/callback/HttpPostClusterCallback.java:
```diff
@@ -0,0 +1,167 @@
+                                     Constants.HttpCallback.DEFAULT_MAX_CONNECTIONS);
+    PoolingHttpClientConnectionManager connectionManager = new PoolingHttpClientConnectionManager();
+    connectionManager.setDefaultMaxPerRoute(maxConnections);
+    connectionManager.setMaxTotal(maxConnections);
```
What happens when maxConnections is reached?

albertshau replied:
blocks on the post call until connections go down.

poornachandra commented on server/src/test/java/com/continuuity/loom/scheduler/SchedulerTest.java:
```diff
@@ -222,6 +227,65 @@ public void testScheduler() throws Exception {
+    clusterScheduler.run();
+
+    // at this point, the start callback should have run, but not the after callbacks
+    Assert.assertEquals(1, mockClusterCallback.getStartCallbacks().size());
```
Since you mentioned that until the onStart callback has returned the job should not start, I think it is a good idea to assert that.

albertshau replied:
done

poornachandra commented on server/src/main/java/com/continuuity/loom/scheduler/callback/HttpPostClusterCallback.java:
```diff
@@ -0,0 +1,167 @@
+ * the HTTP POST request is sent only for specific cluster actions. This is done by specifying a comma separated list
+ * of {@link ClusterAction}s in the configuration for start, success, and/or triggers.
+ */
+public class HttpPostClusterCallback implements ClusterCallback {
```
It would be good to have tests for this class.

albertshau replied:
done

poornachandra commented on server/src/main/java/com/continuuity/loom/scheduler/JobScheduler.java:
```diff
@@ -99,14 +95,27 @@ public void run() {
+          // this can happen if 2 tasks complete around the same time and the first one places the job in the queue,
+          // sees 0 in progress tasks, and sets the cluster status. The job is still in the queue as another element
+          // from the 2nd task and gets here.  In that case, no need to go further.
+          if (job.getJobStatus() == ClusterJob.Status.FAILED && cluster.getStatus() != Cluster.Status.PENDING) {
```
What about setting the task status? Same comment for COMPLETE above.

albertshau replied:
the task itself has already been updated at this point so there is no need.  We add the job onto the queue after the task status is marked as complete in order to check if the stage needs to be advanced of the job itself is done.

albertshau committed:
add missing javadocs

poornachandra commented:
LGTM :+1:
